### ISSUE TITLE
High ATOL in Qwen MLP Linear Layer Validation

### ISSUE THREAD

**Issue by jennychristopher** [tenstorrent/tt-metal#25557]
### Describe the bug

While validating the Linear Layer in the Qwen MLP submodule, we observed the following:
PCC metric: 0.99 
ATOL (Absolute Tolerance): 3.08

We modified the compute kernel configuration for the Linear layer and saw that ATOL dropped from 3.08 to 0.72. The config file change is attached below

However, we would like to understand:
**Is there any approach or recommended configuration to bring the ATOL value to 0?
Given that both reference and TTNN implementations use bfloat16, is it technically feasible to achieve ATOL = 0 in this scenario in all the iterations ?**

Compute Kernel Config Changes:

# Define compute kernel config
        compute_kernel_config = ttnn.WormholeComputeKernelConfig(
            math_fidelity=ttnn.MathFidelity.HiFi4,
            fp32_dest_acc_en=True,
            packer_l1_acc=True
        )
 
        # Linear with GELU activation
        w1_out = ttnn.linear(
            x, self.w1, bias=self.b1, dtype=ttnn.bfloat16, memory_config=ttnn.DRAM_MEMORY_CONFIG, activation="silu",
            compute_kernel_config=compute_kernel_config
        )     

### Steps to reproduce the issue

[1.](url) Go to '...'

**Comment by fahmedchanTT** [comment]
@cmaryanTT This looks related the accuracy push for Ops, can you take a look?

**Comment by cmaryanTT** [comment]
@jennychristopher - can you please provide full steps to reproduce

**Comment by cmaryanTT** [comment]
Atol isn't always a good metric, it doesn't scale with the magnitude of the expected result. allclose is a more reasonable way to look at it, which combines atol and rtol.

atol of zero is essentially impossible. Order of performing operations, differences in rounding and accumulation, approximations for things like activation functions - all of this contributes to accumulated errors.

**Comment by jennychristopher** [comment]
> [@jennychristopher](https://github.com/jennychristopher) - can you please provide full steps to reproduce

**Reproduction Steps:**

Transformers version used: 4.51.3

1. Do retrieve the input pt file from here - https://drive.google.com/drive/folders/1lao2FVcZusXgUfptOspbsVIy_tdRdaOE

2.. In modelling_qwen2_5_vl.py file present in folder tt-metal/python_env/lib/python3.10/site-packages/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py do the following changes 
  

````
class Qwen2_5_VLMLP(nn.Module):
    def __init__(self, config, bias: bool = False):
        super().__init__()
        self.hidden_size = config.hidden_size
        self.intermediate_size = config.intermediate_size
        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=bias)
        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=bias)
        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=bias)
        self.act_fn = ACT2FN[config.hidden_act]
    def forward(self, hidden_state):
        # print(hidden_state.shape)
        # torch.save(self.up_proj.weight, "/home/mlp_weight.pt")
        # torch.save(hidden_state, "/home/vision_mlp.pt")
        # return self.up_proj(hidden_state)
        return self.down_proj(hidden_state)
        # return self.down_proj(self.act_fn(self.gate_proj(hidden_state)) * self.up_proj(hidden_state))`




3. Do checkout to this branch - https://github.com/flexaihq/tt-metal/tree/qwen_atol

4. Execute the following commands to get the metrics atol, rtol, pcc.
export HF_MODEL=Qwen/Qwen2.5-VL-7B-Instruct
pytest tt-metal/models/experimental/qwen_vl/tests/test_mlp.py

**Comment by jennychristopher** [comment]
Kindly let us know the possible workarounds that we can try to minimize the ATOL value further.

**Comment by jennychristopher** [comment]
@cmaryanTT If the math being done on both sides (reference and implementation) is in the same order and same precision why would ATOL not be zero? If zero is impossible, what's the Max ATOL and RTOL we can set? And How we can reduce the ATol and Rtol?

**Comment by cmaryanTT** [comment]
@bbradelTT - by intuition, this feels like MM accuracy. Can we suggest best practices for accuracy?

**Comment by bbradelTT** [comment]
@jennychristopher @cmaryanTT 
what is in the description, setting a high hifi setting, using fp32 dest acc, and packer l1 acc is considered best practice.

The arch independent way of doing that is
```
    compute_kernel_config = ttnn.init_device_compute_kernel_config(
            device.arch(),
            math_fidelity=ttnn.MathFidelity.HiFi4,
            math_approx_mode=False,
            fp32_dest_acc_en=True,
            packer_l1_acc=True,
        )
```

That should reduce atol and rtol as much as possible.

The max ATOL and RTOL would really depend on the inputs. E.g. 0, near 0, and inf values can really cause problems.

There is no simple way to get identical order of operations between torch and ttnn. Our FPU works on very specific block granularities that would be different from what torch does.

**Comment by jennychristopher** [comment]
@bbradelTT  We would like to understand the differences in working of TT FPU Vs torch. Is there a difference in the numerical block format representation in TT vs Torch?

Can you also explain how the order of operations differs between Torch and TTNN and how the FPU block granularity affects numerical differences?

**Comment by bbradelTT** [comment]
Linear may do a matmul followed by an add or might do a matmul fused with an op, and the model probably executes other ops. It's usually good to isolate individual ops to look at accuracy.

I think the order of operations for TTNN matmul/linear at a tile level should be reasonably close to what is described at https://github.com/tenstorrent/tt-isa-documentation/blob/main/WormholeB0/TensixTile/TensixCoprocessor/MVMUL.md

You'd have to combine that with the sequence of tile operations that would depend on which matmul program config is used.

For torch, that will depend on the cpu, libraries, and number of processors (see https://github.com/tenstorrent/tt-metal/issues/24611) and I don't have a link I could point to.

For the data formats, there's various documentation. E.g. 
- general type info https://docs.tenstorrent.com/pybuda/latest/dataformats.html
- data conversion https://github.com/tenstorrent/tt-isa-documentation/blob/main/WormholeB0/TensixTile/TensixCoprocessor/Packers/FormatConversion.md
- has some info on both https://github.com/tenstorrent/tt-isa-documentation/blob/main/WormholeB0/TensixTile/TensixCoprocessor/SrcASrcB.md

**Comment by fahmedchanTT** [comment]
@jennychristopher Do @bbradelTT's comments resolve the initial ask here?

**Comment by fahmedchanTT** [comment]
@jennychristopher  Please re-open if necessary.
